packages/core/src/di/provider_collection.ts

```diff
@@ -19,20 +19,39 @@ import {resolveForwardRef} from './forward_ref';
 import {INJECTOR_INITIALIZER} from './initializer_token';
 import {ɵɵinject as inject} from './injector_compatibility';
 import {getInjectorDef, InjectorType, InjectorTypeWithProviders} from './interface/defs';
-import {ClassProvider, ConstructorProvider, ExistingProvider, FactoryProvider, Provider, StaticClassProvider, TypeProvider, ValueProvider} from './interface/provider';
+import {ClassProvider, ConstructorProvider, ExistingProvider, FactoryProvider, ModuleWithProviders, Provider, StaticClassProvider, TypeProvider, ValueProvider} from './interface/provider';
 import {INJECTOR_DEF_TYPES} from './internal_tokens';
 
+/**
+ * A source of providers for the `importProvidersFrom` function.
+ *
+ * @publicApi
+ */
+export type ImportProvidersSource =
+    Type<unknown>|ModuleWithProviders<unknown>|Array<ImportProvidersSource>;
+
 /**
  * Collects providers from all NgModules and standalone components, including transitively imported
  * ones.
  *
  * @returns The list of collected providers from the specified list of types.
  * @publicApi
  */
-export function importProvidersFrom(...types: Array<Type<unknown>>): Provider[] {
-  const providers: SingleProvider[] = [];
-  deepForEach(types, type => walkProviderTree(type, providers, [], new Set()));
-  return providers;
+export function importProvidersFrom(...sources: ImportProvidersSource[]): Provider[] {
+  const providersOut: SingleProvider[] = [];
+  const injectorTypesWithProviders: InjectorTypeWithProviders<unknown>[] = [];
+  deepForEach(sources, source => {
+    // Narrow `source` to access the internal type analogue for `ModuleWithProviders`.
+    const internalSource = source as Type<unknown>| InjectorTypeWithProviders<unknown>;
+    if (walkProviderTree(internalSource, providersOut, [], new Set())) {
+      injectorTypesWithProviders.push(internalSource);
+    }
+  });
+  for (const {providers} of injectorTypesWithProviders) {
+    providersOut.push(...(providers || EMPTY_ARRAY));
```

Comment:
We should probably handle this in a similar way we do that inside the `walkProviderTree` fn (may be extract this as a separate function to avoid code duplication?):
```suggestion
  for (const {ngModule, providers} of injectorTypesWithProviders) {
    providers ||= EMPTY_ARRAY;
    deepForEach(providers, provider => {
      ngDevMode && validateProvider(provider, providers, ngModule);
      providersOut.push(provider);
    });
```